tests/asgi/test_asgi_servers.py

```diff
@@ -93,6 +96,246 @@ def test_sse(self, server_base_url):
         assert not events[-1]
 
 
+class TestWebSocket:
+
+    @pytest.mark.parametrize('explicit_close', [True, False])
+    @pytest.mark.parametrize('close_code', [None, 4321])
+    def test_hello(self, explicit_close, close_code, server_url_events_ws):
+        echo_expected = 'Check 1 - \U0001f600'
+
+        extra_headers = {'X-Command': 'recv'}
+
+        if explicit_close:
+            extra_headers['X-Close'] = 'True'
+
+        if close_code:
+            extra_headers['X-Close-Code'] = str(close_code)
+
+        async def do_test():
+            async with websockets.connect(
+                server_url_events_ws,
+                extra_headers=extra_headers,
+            ) as ws:
+                got_message = False
+
+                while True:
+                    try:
+                        # TODO: Why is this failing to decode on the other side? (raises an error)
+                        # TODO: Why does this cause Daphne to hang?
```

Comment:
FWIW, I was testing this (`websockets`) module with our uWSGI-based WebSocket (which, FWIW, works flawlessly on all major browsers), and it was hanging in a strange way, particularly if receiving a fair amount of data without sending.